src/asset/asset-list-loader.js

```diff
@@ -0,0 +1,214 @@
+import { EventHandler } from '../core/event-handler.js';
+
+import { Asset } from './asset.js';
+
+/**
+ * @class
+ * @name AssetListLoader
+ * @augments EventHandler
+ * @classdesc Used to load a group of assets and fires a callback when all assets are loaded.
+ * @param {Asset[]|number[]} assetList - An array of {@link Asset} objects to load or an array of Asset IDs to load.
+ * @param {AssetRegistry} assetRegistry - The application's asset registry.
+ */
+class AssetListLoader extends EventHandler {
+    constructor(assetList, assetRegistry) {
+        super();
+        this._assets = [];
+        this._loadingAssets = new Set();
+        this._registry = assetRegistry;
+        this._loaded = false;
+        this._count = 0; // running count of successfully loaded assets
+        this._failed = []; // list of assets that failed to load
+
+        this._waitingAssets = [];
+
+        if (assetList.length && assetList[0] instanceof Asset) {
+            assetList.forEach((asset, i, array) => {
+                // filter out duplicates
+                if (array.indexOf(asset) !== i) {
+                    return;
+                }
+                if (!asset.registry) {
+                    asset.registry = assetRegistry;
+                }
+                this._assets.push(asset);
+            });
+        } else {
+            // list of Asset IDs
+            assetList.forEach((assetId, i, array) => {
+                // filter out duplicates
+                if (array.indexOf(assetId) !== i) {
+                    return;
+                }
+                var asset = assetRegistry.get(assetList[i]);
+                if (asset) {
+                    this._assets.push(asset);
+                } else {
+                    this._waitForAsset(assetList[i]);
+                }
+            });
+        }
+    }
+
+    destroy() {
+        // remove any outstanding listeners
+        var self = this;
+
+        this._registry.off("load", this._onLoad);
+        this._registry.off("error", this._onError);
+
+        this._waitingAssets.forEach(function (id) {
+            self._registry.off("add:" + id, this._onAddAsset);
+        });
+
+        this.off("progress");
+        this.off("load");
+    }
+
+    /**
+     * @function
+     * @name AssetListLoader#load
+     * @description Start loading asset list, call done() when all assets have loaded or failed to load.
+     * @param {Function} done - Callback called when all assets in the list are loaded. Passed (err, failed) where err is the undefined if no errors are encountered and failed contains a list of assets that failed to load.
+     * @param {object} [scope] - Scope to use when calling callback.
+     *
+     */
+    load(done, scope) {
+        var i = 0;
+        var l = this._assets.length;
+        var asset;
```

Comment:
We don't use `var` any more in engine source (except JSDoc comments - although even that should go soon).

Reply:
There are several other instances in this file. Please migrate them all. There is an ESLint rule that prevents use of `var` which I want to turn on but we need to upgrade the JSDoc occurrences first.